Alternate vCPU suspension method

bsdjhb opened the pull request:
This compiles, but I haven't run-time tested it. :(  This is what I tried to describe on the phone earlier today about using the debug vmexit as the hook for synchronizing with the checkpoint thread.  In this case I went ahead and left the thread as-is rather than being driven from the event of the last vCPU suspending.  It was a smaller patch and also avoids silliness with the case where all the vCPUs are already idle, etc.

bsdjhb committed:
Use debug VM exits to pause vCPUs during a snapshot.

- Leave the existing logic to suspend vCPUs as-is and modify
  the vmexit_debug handler in bhyve to notice when vCPUs are
  idle.
- Move vm_vcpu_suspend/resume into bhyve instead of the library
  and have them wait for all vCPUso to go idle as reported
  by debug VM exits.
- To compose with the debug server invoke one callback when a
  vCPU exits from the kernel and is idle and a second one before
  the vCPU returns back to the kernel.

bsdjhb committed:
Also mark vCPUs in an MTRAP exit as idle for checkpoints.

Most VM exit types will return to the kernel "soon" and so it is ok to
wait for them to do that and pop back out with a debug VM exit.  However,
mtrap exits might result in the vCPU thread waiting in the debug server
for an indeterminate amount of time, so handle checkpointing for mtrap
exits as well.

elena19m commented on usr.sbin/bhyve/bhyverun.c:
```diff
@@ -710,7 +711,9 @@ vmexit_mtrap(struct vmctx *ctx, struct vm_exit *vmexit, int *pvcpu)
 
 	stats.vmexit_mtrap++;
 
+	checkpoint_cpu_suspend(*pvcpu);
```
Is there any reason to add `checkpoint_cpu_suspend()` and `checkpoint_cpu_resume()` in `vmexit_mtrap`? We don't use `VM_CAP_MTRAP_EXIT` in our code and I don't know if these changes produce any effect or not. Are these changes made in order for gdb to work along with the snapshot process?

bsdjhb replied:
Sorry, I tried to explain this in the second commit log message.  When a vCPU is set in 'suspend_cpus', any attempt to run in the guest will immediately exit with VM_EXITCODE_DEBUG.  Thus, if a vCPU thread is in userland handling an exit like an access to a virtual device, it will eventually pop back out and be noticed by the vmexit_debug handler.  However, that only works for VM exits that immediately don't sit in userland and wait.  This is true for most VM exits but it is not true of the VM exits used by the debug server (gdb) as the vCPUs might sit paused in the debug server until the remote debugger sends some kind of command.  Right now there are only two such events, vmexit_debug if the debugger suspends vCPUs, and vmexit_mtrap when single-stepping.  But, yes, this is only needed for the debug server.  You could test this change by attaching to the VM with gdb and single-stepping one of the vCPUs via a 's' or 'si' command in gdb, then taking a snapshot while the guest is paused in the debugger.  Without this change the snapshot would hang as the stepping vCPU would never look like it was idle.  With this change the snapshot would work.

bsdjhb replied:
Sorry, 'debug_cpus', not 'suspend_cpus' above.

elena19m commented on usr.sbin/bhyve/snapshot.c:
```diff
@@ -971,6 +976,76 @@ vm_mem_write_to_file(int fd, const void *src, size_t dst_offset, size_t len)
+
+	pthread_mutex_lock(&vcpu_lock);
+	checkpoint_active = true;
+	vm_suspend_cpu(ctx, -1);
```
While looking for a way to force the vcpus to return to the userspace, the `vm_suspend_cpu` function was one of the easiest way through which we could call `vcpu_notify_event` to a vcpu. However, we don't know if it is OK to add the vcpus in the `vm->debug_cpus` set. That's why we changed the implementation and added that `no_debug` additional paramenter.

bsdjhb replied:
It is ok to set the vCPU in `debug_cpus`.  That ensures that the vCPU will eventually exit into vmexit_debug().  This version is using `debug_cpus` to replace libvmmapi changes in the current version (more or less).  It should be a bit cheaper (especially in the common case when a checkpoint hasn't been requested) and avoids the vCPU threads contending on a global lock.  It also permits using checkpoint while the debug server is active.

elena19m commented:
We have tested the patch. It seems to work fine for Linux, Windows and FreeBSD guests.

bsdjhb commented on usr.sbin/bhyve/snapshot.c:
```diff
@@ -971,6 +976,76 @@ vm_mem_write_to_file(int fd, const void *src, size_t dst_offset, size_t len)
+{
+
+	pthread_mutex_lock(&vcpu_lock);
+	checkpoint_active = false;
```
I have a bug here I need to fix.  This needs to call vm_resume_cpu(ctx, -1) to release all the vCPUs so that they can resume execution once the checkpoint has completed.  I will push an update to fix that in a bit.

bsdjhb committed:
Resume all the vCPUs when completing a checkpoint.

Without this, vCPUs would keep exiting with VM_EXITCODE_DEBUG
endlessly after a checkpoing was completed.
